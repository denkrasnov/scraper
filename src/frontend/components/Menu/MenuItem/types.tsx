import { DataAttr } from "~app/common/types";

export interface MenuItemProps {
  src: string;
  text: string;
  onClick: () => void;
  dataAttr?: DataAttr;
}
