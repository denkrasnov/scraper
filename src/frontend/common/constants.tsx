import { Channel } from "../../backend/scrapers/types";

export const channelLogo: Record<string, string> = {
  [Channel.JurnalTV]:
    "https://live-tv-channels.org/pt-data/uploads/logo/md-jurnal-tv.jpg",
  [Channel.NTV]:
    "https://yt3.ggpht.com/ytc/AAUvwnjTxNZwMTMTlj18m75pswpomJFvJEHuargB5gWA=s900-c-k-c0x00ffffff-no-rj",
  [Channel.TV8]: "https://tv8.md/wp-content/uploads/2020/07/logo-tv8-new-2.png"
};
