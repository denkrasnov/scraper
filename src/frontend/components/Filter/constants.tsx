// TODO: created shared types with /backend
import { ProductName } from "../../../backend/scrapers/types";

export const filterName = {
  [ProductName.MD]: "Canal"
};
