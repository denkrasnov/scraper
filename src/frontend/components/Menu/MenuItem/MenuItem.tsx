import React, { FC } from "react";

import Box from "~app/atoms/Box";
import Image from "~app/atoms/Image";
import { H3 } from "~app/atoms/Headings";
import { MenuItemProps } from "./types";

const MenuItem: FC<MenuItemProps> = (props) => {
  const { dataAttr, src, text, onClick } = props;

  return (
    <Box
      cursor="pointer"
      dataAttr={dataAttr}
      flexDirection="column"
      marginRight="s48"
      onClick={onClick}
      width="200px"
    >
      <Image alt="menu-image" height="134px" src={src} />
      <Box justifyContent="center" width="100%">
        <H3>{text}</H3>
      </Box>
    </Box>
  );
};

export default MenuItem;
