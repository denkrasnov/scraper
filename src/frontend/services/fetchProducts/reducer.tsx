import { Action, State, ActionTypes } from "./types";

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case ActionTypes.FETCH:
      return {
        ...state,
        query: action.payload,
        isError: ""
      };
    case ActionTypes.FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: ""
      };
    case ActionTypes.FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        products: action.payload
      };
    case ActionTypes.FETCH_ERROR:
      return {
        ...state,
        isLoading: false,
        isError: action.payload
      };

    default:
      throw new Error();
  }
};

export default reducer;
