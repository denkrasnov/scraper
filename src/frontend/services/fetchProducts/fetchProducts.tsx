import { useEffect, useReducer, Dispatch } from "react";
import axios from "axios";

import { INITIAL_STATE } from "./constants";
import reducer from "./reducer";
import { Action, State, ActionTypes, Response } from "./types";

const fetchProducts = (): [State, Dispatch<Action>] => {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  const { query } = state;

  useEffect(() => {
    let didCancel = false;

    const fetchData = async () => {
      dispatch({ type: ActionTypes.FETCH_INIT });
      try {
        const result: Response = await axios.get(`/search/${query}`);

        if (!didCancel) {
          dispatch({
            type: ActionTypes.FETCH_SUCCESS,
            payload: result.data.products
          });
        }
      } catch (error) {
        if (!didCancel) {
          dispatch({ type: ActionTypes.FETCH_ERROR, payload: error });
        }
      }
    };

    if (query) {
      fetchData();
    }

    return () => {
      didCancel = true;
    };
  }, [query]);

  return [state, dispatch];
};

export default fetchProducts;
