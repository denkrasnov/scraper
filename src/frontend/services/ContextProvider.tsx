import React, { FC, useContext, createContext } from "react";

import { INITIAL_STATE } from "./fetchProducts/constants";
import fetchProducts from "./fetchProducts";
import { ContextProviderProps, CreateContext } from "./types";

export const StateContext = createContext<CreateContext>([
  INITIAL_STATE,
  () => {}
]);

const ContextProvider: FC<ContextProviderProps> = ({ children }) => (
  <StateContext.Provider value={fetchProducts()}>
    {children}
  </StateContext.Provider>
);

export const useFullContext = () => useContext(StateContext);

export default ContextProvider;
