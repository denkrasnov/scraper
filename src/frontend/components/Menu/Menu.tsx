import React from "react";

import Box from "~app/atoms/Box";
import { useFullContext } from "~app/services/ContextProvider";
import { ActionTypes } from "~app/services/fetchProducts/types";
import { image } from "~app/assets";
import MenuItem from "./MenuItem";

const Menu = () => {
  const [, dispatch] = useFullContext();

  return (
    <Box
      justifyContent="center"
      marginBottom="s32"
      marginTop="s20"
      width="100%"
    >
      <MenuItem
        dataAttr={{ "data-e2e-id": "menuItem.tv" }}
        onClick={() => dispatch({ type: ActionTypes.FETCH, payload: "tvs" })}
        src={image.tv}
        text="Телевизоры"
      />
      <MenuItem
        dataAttr={{ "data-e2e-id": "menuItem.fridge" }}
        onClick={() =>
          dispatch({ type: ActionTypes.FETCH, payload: "fridges" })
        }
        src={image.fridge}
        text="Холодильники"
      />
    </Box>
  );
};

export default Menu;
