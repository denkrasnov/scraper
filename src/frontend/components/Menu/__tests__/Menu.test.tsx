import React from "react";
import { shallow } from "enzyme";

import * as AppContext from "~app/services/ContextProvider";
import { INITIAL_STATE } from "~app/services/fetchProducts/constants";
import MenuItem from "../MenuItem";
import Menu from "..";

describe("Menu", () => {
  it("should render", () => {
    jest
      .spyOn(AppContext, "useFullContext")
      .mockImplementation(() => [INITIAL_STATE, () => {}]);

    const component = shallow(<Menu />);

    expect(component).toMatchSnapshot();
  });

  it("should call onClick and dispatch", () => {
    const dispatchMock = jest.fn();
    jest
      .spyOn(AppContext, "useFullContext")
      .mockImplementation(() => [INITIAL_STATE, dispatchMock]);

    const component = shallow(<Menu />);

    component.find(MenuItem).at(0).simulate("click");
    expect(dispatchMock).toHaveBeenCalledTimes(1);
  });

  it("should call onClick and dispatch on second MenuItem", () => {
    const dispatchMock = jest.fn();
    jest
      .spyOn(AppContext, "useFullContext")
      .mockImplementation(() => [INITIAL_STATE, dispatchMock]);

    const component = shallow(<Menu />);

    component.find(MenuItem).at(1).simulate("click");
    expect(dispatchMock).toHaveBeenCalledTimes(1);
  });
});
