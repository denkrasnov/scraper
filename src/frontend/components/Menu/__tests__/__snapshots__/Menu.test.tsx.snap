// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render 1`] = `
<Box
  justifyContent="center"
  marginBottom="s32"
  marginTop="s20"
  width="100%"
>
  <MenuItem
    dataAttr={
      Object {
        "data-e2e-id": "menuItem.tv",
      }
    }
    onClick={[Function]}
    src="test-file-stub"
    text="Телевизоры"
  />
  <MenuItem
    dataAttr={
      Object {
        "data-e2e-id": "menuItem.fridge",
      }
    }
    onClick={[Function]}
    src="test-file-stub"
    text="Холодильники"
  />
</Box>
`;
